File: 244__Shortest_Word_Distance_II.py
class WordDistance(object):

    def __init__(self, words):
        """
        :type words: List[str]
        """
        self.dic = collections.defaultdict(list)
        for i, word in enumerate(words):
            self.dic[word].append(i)
        self.max = len(words)


    def shortest(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        i1_list = self.dic[word1]
        i2_list = self.dic[word2]

        i1 = 0
        i2 = 0
        result = self.max

        while i1 < len(i1_list) and i2 < len(i2_list):
            if i1_list[i1] > i2_list[i2]:
                result = min(i1_list[i1] - i2_list[i2], result)
                i2 += 1
            else:
                result = min(i2_list[i2] - i1_list[i1], result)
                i1 += 1
        return result



# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(words)
# param_1 = obj.shortest(word1,word2)

File: Java/244__Shortest_Word_Distance_II.java
class WordDistance {

    HashMap<String, ArrayList<Integer>> locations;

    public WordDistance(String[] words) {
        this.locations = new HashMap<>();
        for (int i = 0; i < words.length; i++) {
            if (!locations.containsKey(words[i])) {
                locations.put(words[i], new ArrayList<Integer>());
            }
            locations.get(words[i]).add(i);
        }
    }

    public int shortest(String word1, String word2) {
        ArrayList<Integer> map_1 = this.locations.get(word1), map_2 = this.locations.get(word2);
        int i = 0, j = 0, result = Integer.MAX_VALUE;
        while (i < map_1.size() && j < map_2.size()) {
            result = Math.min(result, Math.abs(map_1.get(i) - map_2.get(j)));
            if (map_1.get(i) < map_2.get(j)) {
                i++;
            } else {
                j++;
            }
        }
        return result;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(words);
 * int param_1 = obj.shortest(word1,word2);
 */
